File: .github/validate_preview.py
import json, os, re, sys
from urllib.parse import urljoin
import requests
from bs4 import BeautifulSoup

BASE = os.environ['PREVIEW_BASE'].rstrip('/')
CANON = 'https://osaraclinics.com'
PATHS = ['/', '/dermatology', '/ophthalmology', '/doctors/dr-osama-alwreikat', '/doctors/dr-sara-abu-touq', '/vitiligo-jordan', '/botox-hyperhidrosis', '/mole-removal', '/school-health']
EXPECTED_HOURS = [
    {'dayOfWeek': ['Saturday','Sunday'], 'opens':'16:30', 'closes':'20:30'},
    {'dayOfWeek': ['Monday','Tuesday','Wednesday','Thursday'], 'opens':'08:00', 'closes':'13:30'},
    {'dayOfWeek': ['Monday','Tuesday','Wednesday','Thursday'], 'opens':'16:30', 'closes':'20:30'},
]
s = requests.Session(); s.headers['User-Agent']='OSara-PR1-final-validator/1.0'
errors=[]; pages=[]; links=set()

def get(path):
    return s.get(BASE + path, allow_redirects=True, timeout=30)

for path in PATHS:
    r=get(path)
    soup=BeautifulSoup(r.text,'html.parser')
    canonical=soup.find('link', rel=lambda x:x and 'canonical' in x)
    h1s=soup.find_all('h1')
    robots=[m.get('content','') for m in soup.find_all('meta', attrs={'name':re.compile('^robots$',re.I)})]
    expected=CANON+'/' if path=='/' else CANON+path
    row={'path':path,'status':r.status_code,'redirects':len(r.history),'canonical':canonical.get('href') if canonical else None,'h1_count':len(h1s),'title':soup.title.get_text(' ',strip=True) if soup.title else None,'noindex':any('noindex' in x.lower() for x in robots)}
    pages.append(row)
    if r.status_code!=200 or len(r.history)!=0: errors.append(f'{path}: expected direct 200')
    if row['canonical']!=expected: errors.append(f'{path}: canonical {row["canonical"]} != {expected}')
    if len(h1s)!=1: errors.append(f'{path}: H1 count {len(h1s)}')
    if not row['title']: errors.append(f'{path}: missing title')
    if row['noindex']: errors.append(f'{path}: accidental noindex')
    for a in soup.find_all('a',href=True):
        href=a['href'].split('#')[0]
        if href.startswith('/') and href:
            links.add(href)

broken=[]
for href in sorted(links):
    r=get(href)
    if r.status_code>=400:
        broken.append({'href':href,'status':r.status_code})
if broken: errors.append('broken internal links: '+json.dumps(broken))

robots=get('/robots.txt'); sitemap=get('/sitemap.xml')
if robots.status_code!=200 or 'Sitemap: https://osaraclinics.com/sitemap.xml' not in robots.text: errors.append('robots invalid')
if sitemap.status_code!=200: errors.append('sitemap not 200')
sitemap_urls=re.findall(r'<loc>(.*?)</loc>',sitemap.text)
expected_urls=[CANON+'/' if p=='/' else CANON+p for p in PATHS]
if sitemap_urls!=expected_urls: errors.append('sitemap URLs mismatch')

home=BeautifulSoup(get('/').text,'html.parser')
clinic=None
for sc in home.find_all('script',attrs={'type':'application/ld+json'}):
    try:
        obj=json.loads(sc.string or sc.get_text())
    except Exception:
        continue
    if isinstance(obj,dict) and obj.get('@type')=='MedicalClinic' and obj.get('@id')==CANON+'/#clinic': clinic=obj
if not clinic: errors.append('MedicalClinic JSON-LD missing')
else:
    actual=[{'dayOfWeek':x.get('dayOfWeek'),'opens':x.get('opens'),'closes':x.get('closes')} for x in clinic.get('openingHoursSpecification',[])]
    if actual!=EXPECTED_HOURS: errors.append('openingHoursSpecification mismatch: '+json.dumps(actual))
    if any('Friday' in (x.get('dayOfWeek') if isinstance(x.get('dayOfWeek'),list) else [x.get('dayOfWeek')]) for x in clinic.get('openingHoursSpecification',[])): errors.append('Friday incorrectly has opening hours')

body=home.get_text(' ',strip=True)
for required in ['Saturday - Sunday','Monday - Thursday','Friday','Closed','Appointments and availability may occasionally vary','For urgent dermatological or ophthalmological concerns outside regular clinic hours','nearest appropriate emergency department']:
    if required not in body: errors.append('visible hours/contact missing: '+required)
if 'Call to enquire' in body or 'Hours vary - Call to confirm' in body: errors.append('old variable-hours wording still visible')
if re.search(r'24\s*/\s*7|24\s*hours', body, re.I): errors.append('24/7 implication found')

for path,bad in [('/doctors/dr-osama-alwreikat','dr_osama.png'),('/doctors/dr-sara-abu-touq','dr_sara.png')]:
    text=get(path).text
    soup=BeautifulSoup(text,'html.parser')
    if bad in text: errors.append(path+': mismatched asset reference remains')
    if soup.find('meta',attrs={'property':'og:image'}): errors.append(path+': OG image remains')
    if soup.find('img'): errors.append(path+': visible image remains')
    for sc in soup.find_all('script',attrs={'type':'application/ld+json'}):
        try: obj=json.loads(sc.string or sc.get_text())
        except Exception: continue
        if '"image"' in json.dumps(obj): errors.append(path+': schema image remains')

report={'pages':pages,'broken_internal_links':broken,'robots_status':robots.status_code,'sitemap_status':sitemap.status_code,'sitemap_urls':sitemap_urls,'hours':clinic.get('openingHoursSpecification',[]) if clinic else None,'errors':errors}
print(json.dumps(report,ensure_ascii=False,indent=2))
sys.exit(1 if errors else 0)

File: .github/validate_preview.mjs
import { chromium } from 'playwright';
import fs from 'fs';

const base=process.env.PREVIEW_BASE.replace(/\/$/,'');
const browser=await chromium.launch({headless:true});
const mobilePaths=['/','/dermatology','/ophthalmology','/doctors/dr-osama-alwreikat','/doctors/dr-sara-abu-touq','/vitiligo-jordan'];
const mobile=[]; const errors=[];
for (const p of mobilePaths) {
  const ctx=await browser.newContext({viewport:{width:390,height:844},deviceScaleFactor:1});
  const page=await ctx.newPage(); const errs=[];
  page.on('pageerror',e=>errs.push('pageerror: '+String(e)));
  page.on('console',m=>{if(m.type()==='error') errs.push('console: '+m.text())});
  const response=await page.goto(base+p,{waitUntil:'networkidle',timeout:45000});
  const metrics=await page.evaluate(()=>({innerWidth:innerWidth,scrollWidth:document.documentElement.scrollWidth,bodyScrollWidth:document.body.scrollWidth,h1Count:document.querySelectorAll('h1').length,visibleH1:[...document.querySelectorAll('h1')].filter(e=>{const r=e.getBoundingClientRect(),s=getComputedStyle(e);return r.width>0&&r.height>0&&s.display!=='none'&&s.visibility!=='hidden'}).length,brokenImages:[...document.images].filter(i=>i.complete&&i.naturalWidth===0).map(i=>i.src),canonical:document.querySelector('link[rel="canonical"]')?.href||null}));
  const name=p==='/'?'home':p.slice(1).replaceAll('/','_');
  await page.screenshot({path:`mobile-${name}.png`,fullPage:true});
  mobile.push({path:p,status:response?.status(),url:page.url(),...metrics,consoleErrors:errs});
  if(response?.status()!==200||metrics.scrollWidth>390||metrics.bodyScrollWidth>390||metrics.h1Count!==1||metrics.visibleH1!==1||metrics.brokenImages.length||errs.length) errors.push('mobile/render '+p);
  await ctx.close();
}

const ctx=await browser.newContext({viewport:{width:1280,height:900}});
const page=await ctx.newPage(); const consoleErrors=[];
page.on('pageerror',e=>consoleErrors.push('pageerror: '+String(e)));
page.on('console',m=>{if(m.type()==='error') consoleErrors.push('console: '+m.text())});
await page.addInitScript(()=>document.addEventListener('click',e=>{if(e.target.closest?.('a[href]'))e.preventDefault()},true));
await page.goto(base+'/',{waitUntil:'networkidle',timeout:45000});

const rendered=await page.evaluate(()=>{
  const body=document.body.innerText;
  let clinic=null;
  for(const s of document.querySelectorAll('script[type="application/ld+json"]')){try{const x=JSON.parse(s.textContent);if(x?.['@type']==='MedicalClinic'&&x?.['@id']==='https://osaraclinics.com/#clinic')clinic=x}catch{}}
  return {hoursText:document.querySelector('.hours-container')?.innerText||'',hasUrgentArabic:body.includes('تواصل للحالات العاجلة خارج ساعات العمل'),hasUrgentEnglish:body.includes('For urgent dermatological or ophthalmological concerns outside regular clinic hours'),has24:/24\s*\/\s*7|24\s*hours/i.test(body),clinicHours:clinic?.openingHoursSpecification||null};
});
const expectedHours=[
 {'@type':'OpeningHoursSpecification',dayOfWeek:['Saturday','Sunday'],opens:'16:30',closes:'20:30'},
 {'@type':'OpeningHoursSpecification',dayOfWeek:['Monday','Tuesday','Wednesday','Thursday'],opens:'08:00',closes:'13:30'},
 {'@type':'OpeningHoursSpecification',dayOfWeek:['Monday','Tuesday','Wednesday','Thursday'],opens:'16:30',closes:'20:30'}
];
if(JSON.stringify(rendered.clinicHours)!==JSON.stringify(expectedHours)) errors.push('rendered schema hours mismatch');
for(const x of ['Saturday - Sunday','Monday - Thursday','Friday','Closed','8:00 AM - 1:30 PM','4:30 PM - 8:30 PM']) if(!rendered.hoursText.includes(x)) errors.push('rendered visible hours missing '+x);
if(!rendered.hasUrgentArabic||!rendered.hasUrgentEnglish||rendered.has24) errors.push('urgent contact messaging invalid');

async function testCTA(path,selector,eventName){
  await page.goto(base+path,{waitUntil:'networkidle',timeout:45000});
  await page.evaluate(()=>{window.gtag=undefined;window.dataLayer=[]});
  const result=await page.evaluate(({selector,eventName})=>{
    const a=document.querySelector(selector); if(!a)return {found:false,eventName};
    const href=a.getAttribute('href'); a.click(); const dl=window.dataLayer||[];
    return {found:true,href,eventName,eventCount:dl.filter(x=>x?.event===eventName).length,dataLayer:dl,forbiddenKeys:[...new Set(dl.flatMap(x=>Object.keys(x||{})).filter(k=>/patient|diagnosis|symptom|message|text/i.test(k)))]};
  },{selector,eventName});
  if(!result.found||result.eventCount!==1||result.forbiddenKeys.length) errors.push('CTA '+eventName);
  return result;
}
const ctas=[];
ctas.push(await testCTA('/','a[href*="wa.me/"]:not([data-appointment]):not(.book-appointment)','whatsapp_click'));
ctas.push(await testCTA('/','a[href^="tel:"]','phone_click'));
ctas.push(await testCTA('/dermatology','a[href*="google.com/maps"]','directions_click'));
ctas.push(await testCTA('/','a[data-appointment],a.book-appointment','appointment_click'));
if(consoleErrors.length) errors.push('desktop console errors');

const report={mobile,rendered,ctas,consoleErrors,errors};
fs.writeFileSync('browser-report.json',JSON.stringify(report,null,2));
console.log(JSON.stringify(report,null,2));
await browser.close();
if(errors.length) process.exit(1);
